Some minor improvements to FakeTensor testing

zou3519 opened the pull request:
Summary:
- PyTorch testing chokes sometimes when it sees an exception where the first
  argument is not a string. fake_tensor.UnsupportedOperatorException's first
  arg is an OpOverload. This PR fixes PyTorch testing to not choke. I'm not
  really sure how to reproduce this in OSS.
- It turns out that if an operator does not have a meta kernel, the FakeTensor
  rule is really slow (30ms in OSS in debug mode, 3s on some internal config).
  The thing that is slow (aside from the previous diff) is waiting for the Dispatcher to
  report NotImplemented and then attempting to catch that. I'm not really sure
  why this is slow but it's easy to workaround so I added a workaround.

Test Plan: - existing tests

Differential Revision: D47917554

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1462,16 +1462,29 @@ def can_run_unsafe_fallback(func: OpOverload):
+
+        # Optimization: If there is no Meta kernel, it takes a surprisingly long
+        # amount of time to catch the NotImplementedError, so we check it here.
+        if not torch._C._dispatch_has_computed_kernel_for_dispatch_key(
+            func.name(), "Meta"
+        ):
```
Just to confirm - will this api work for all of the following meta registrations ? `python meta registrations, prims, decomps, and c++ meta fns`

zou3519 replied:
Yes. All of the listed meta registrations end up registering a implementation to the PyTorch dispatcher at DispatchKey::Meta. This API checks if there is an implementation in the PyTorch dispatcher that may be used for DispatchKey::Meta

zou3519 committed:
Some minor improvements to FakeTensor testing (#106311)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106311

- PyTorch testing chokes sometimes when it sees an exception where the first
  argument is not a string. fake_tensor.UnsupportedOperatorException's first
  arg is an OpOverload. This PR fixes PyTorch testing to not choke. I'm not
  really sure how to reproduce this in OSS.
- It turns out that if an operator does not have a meta kernel, the FakeTensor
  rule is really slow (30ms in OSS in debug mode, 3s on some internal config).
  The thing that is slow (aside from the previous diff) is waiting for the Dispatcher to
  report NotImplemented and then attempting to catch that. I'm not really sure
  why this is slow but it's easy to workaround so I added a workaround.

Test Plan: - existing tests

Reviewed By: eellison

Differential Revision: D47917554

fbshipit-source-id: f5efefea6031e2c6ead8361db12ba240f8282a67
